moquette-0.16/broker/src/main/java/io/moquette/interception/BrokerInterceptor.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * The Eclipse Public License is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * The Apache License v2.0 is available at
+ * http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.moquette.interception;
+
+import io.moquette.BrokerConstants;
+import io.moquette.interception.messages.*;
+import io.moquette.broker.config.IConfig;
+import io.moquette.broker.subscriptions.Subscription;
+import io.netty.handler.codec.mqtt.MqttConnectMessage;
+import io.netty.handler.codec.mqtt.MqttPublishMessage;
+import io.netty.util.ReferenceCountUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import static io.moquette.logging.LoggingUtils.getInterceptorIds;
+
+/**
+ * An interceptor that execute the interception tasks asynchronously.
+ */
+public final class BrokerInterceptor implements Interceptor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BrokerInterceptor.class);
+    private final Map<Class<?>, List<InterceptHandler>> handlers;
+    private final ExecutorService executor;
+
+    private BrokerInterceptor(int poolSize, List<InterceptHandler> handlers) {
+        LOG.info("Initializing broker interceptor. InterceptorIds={}", getInterceptorIds(handlers));
+        this.handlers = new HashMap<>();
+        for (Class<?> messageType : InterceptHandler.ALL_MESSAGE_TYPES) {
+            this.handlers.put(messageType, new CopyOnWriteArrayList<InterceptHandler>());
+        }
+        for (InterceptHandler handler : handlers) {
+            this.addInterceptHandler(handler);
+        }
+        executor = Executors.newFixedThreadPool(poolSize);
+    }
+
+    /**
+     * Configures a broker interceptor, with a thread pool of one thread.
+     *
+     * @param handlers InterceptHandlers listeners.
+     */
+    public BrokerInterceptor(List<InterceptHandler> handlers) {
+        this(1, handlers);
+    }
+
+    /**
+     * Configures a broker interceptor using the pool size specified in the IConfig argument.
+     * @param props configuration properties.
+     * @param handlers InterceptHandlers listeners.
+     *
+     */
+    public BrokerInterceptor(IConfig props, List<InterceptHandler> handlers) {
+        this(Integer.parseInt(props.getProperty(BrokerConstants.BROKER_INTERCEPTOR_THREAD_POOL_SIZE, "1")), handlers);
+    }
+
+    /**
+     * Shutdown graciously the executor service
+     */
+    public void stop() {
+        LOG.info("Shutting down interceptor thread pool...");
+        executor.shutdown();
+        try {
+            LOG.info("Waiting for thread pool tasks to terminate...");
+            executor.awaitTermination(10L, TimeUnit.SECONDS);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Check the return value of `awaitTermination()` to determine if the operation timed out while waiting for other threads to stop executing. Handle when `awaitTermination()` times out. See the following diffs on GitHub that make this code change: [diff1](https://github.com/voldemort/voldemort/commit/e44ec88c57c1293a5b71598e0ad0fed403387744#diff-c220173cebaeedd4ef7bad06e679a0938710546ed77afac1d37a3ddf25e30e97L533), [diff2](https://github.com/utPLSQL/utPLSQL-cli/commit/079e021a747fe3829f868ad08c95a6eef2cc2f1c#diff-8a4bc1fbdd325f803130235d0cc70d6ae0d4b491b7440263c4f8a55452fd45eeL204)

Learn more: [ExecutorService.awaitTermination in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#awaitTermination-long-java.util.concurrent.TimeUnit-)